File: tasks/2023/final/div3.cpp
// linear, greedy
// @check-accepted: *

#include <array>
#include <iostream>
#include <vector>

using namespace std;

int mod3(const string& s) {
    int sum = 0;
    for (char c : s) sum += c - '0';
    return sum % 3;
}

string without_leading_0(const string& s) {
    for (int i = 0; i < s.size(); i++) {
        if (s[i] != '0') return s.substr(i);
    }
    return "";
}

// biggest result if cnt digits of rem remainder are deleted
string remove(string s, int cnt, int rem) {
    // remove first where the next is bigger
    for (int i = 0; i + 1 < s.size(); i++) {
        if ((s[i] - '0') % 3 == rem && s[i + 1] > s[i]) {
            s.erase(i, 1);
            i = max(-1, i - 2);
            cnt--;
            if (cnt == 0) return s;
        }
    }
    // remove last (next will be smaller or equal for sure)
    for (int i = s.size() - 1; i >= 0; i--) {
        if ((s[i] - '0') % 3 == rem) {
            s.erase(i, 1);
            cnt--;
            if (cnt == 0) return s;
        }
    }
    return "";
}

void solve() {
    string s;
    cin >> s;
    int rem = mod3(s);
    if (rem == 0) {
        cout << s << "\n";
        return;
    }
    string one = without_leading_0(remove(s, 1, rem));
    string two = without_leading_0(remove(s, 2, 3 - rem));
    if (one.size() > two.size() || one.size() == two.size() && one > two) {
        cout << one << "\n";
    } else {
        if (two != "") cout << two << "\n";
        else cout << "-1\n";
    }
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);
    int N;
    cin >> N;
    for (int i = 0; i < N; i++) {
        solve();
    }

    return 0;
}

File: tasks/2023/final/div3.py
#!/usr/bin/env python3
# @check-accepted: *

# biggest result if cnt digits of rem remainder are deleted
def remove(s, cnt, rem):
    i = 0
    while i + 1 < len(s):
        if s[i] % 3 == rem and s[i] < s[i + 1]:
            s = s[: i] + s[i + 1 :]
            cnt -= 1
            if cnt == 0:
                return s
            if i > 0: i -= 1
        else:
            i += 1
    i = len(s) - 1
    while i >= 0:
        if s[i] % 3 == rem:
            s = s[: i] + s[i + 1 :]
            cnt -= 1
            if cnt == 0:
                return s
        i -= 1
    return []

def without_leading_0(s):
    for i in range(len(s)):
        if s[i] > 0:
            return s[i:]
    return []

def solve():
    s = [int(c) for c in input()]
    rem = sum(s) % 3
    if rem == 0:
        print(''.join([str(x) for x in s]))
        return
    one = without_leading_0(remove(s.copy(), 1, rem))
    two = without_leading_0(remove(s.copy(), 2, 3 - rem))
    if len(one) > len(two) or len(one) == len(two) and one > two:
        print(''.join([str(x) for x in one]))
    elif len(two) > 0:
        print(''.join([str(x) for x in two]))
    else:
        print(-1)

N = int(input().strip())
for _ in range(N):
    solve()
